### DISCUSSION TITLE
Is there a csv for all weather station metadata?

### DISCUSSION THREAD

**Discussion by DrewScatterday** [weather-gov/api#721] [Q&A] [answered]
The endpoint /stations maxes out at a 500 limit. So if I wanted metadata about all the weather stations I would have to send a lot of requests because I presume there is several thousand total weather stations. Is there a csv or xml file hosted somewhere that gives all metadata (lat, long, name, ...etc) for each station? 

For example NDBC has something like this with the buoy stations https://www.ndbc.noaa.gov/activestations.xml

What I'm trying to do is find a list of metadata of wind stations 

Thanks for your time! 
-Drew

**Comment by ktrue** [comment]
You might try the list of METAR stations from

https://aviationweather.gov/data/cache/stations.cache.xml.gz

as a list you can use.

I do note that their server seems to have an invalid GZ format when obtained by PHP so I use code
```
$goodHeader = "\x1f\x8b\x08\x08"; # 4-byte GZ file header
if(substr($rawGZ,0,4) === $goodHeader) {# check for valid GZ header
  $Debug .= "..GZ file has good header\n";
	$rawXML = gzdecode($rawGZ);
} else {
  $Debug .= "--GZ file has missing header .. prepending it.\n";
  $rawXML = gzdecode($goodHeader.$rawGZ);
}
```
to fix the 'pull' and successfully gzdecode() the result.

See the full code I use in https://github.com/ktrue/metar-placefile in the get-aviation-metadata.php script.

**Comment by DrewScatterday** [comment, marked as answer]
What I ended up doing is using the /stations endpoint with pagination, although the pagination cursor seems bugged to me but maybe I'm missing something. The loop would go on infinitely and the cursor would never run out/change to null

Ultimately I just hacked around it with using a counter. There are 45832 stations, if we can get 500 every time then we need 91 iterations to get all stations. I then loaded them into a pandas dataframe which then allows me to do some of the analysis I wanted to do with the metadata. Would be great if there was a csv or xml file somewhere so one does not have to make 91 individual requests to get all the data but not a huge deal 

```python
import requests
import pandas as pd 

def get_geojson_data(url):
    try:
        response = requests.get(url)
        response.raise_for_status() 
        if response.headers.get('content-type') == 'application/geo+json':
            geojson_data = response.json()
            return geojson_data
        else:
            print("Response is not in GeoJSON format")
            return None
    except requests.exceptions.RequestException as e:
        print("Error fetching data:", e)
        return None

all_data = []
next_url = "https://api.weather.gov/stations"
count = 0
while next_url: 
    print(next_url)
    data = get_geojson_data(next_url)
    for item in data["features"]:
        coords = item["geometry"]["coordinates"]
        all_data.append([item["id"], coords[0], coords[1]])

    pagination = data.get('pagination')
    if pagination and 'next' in pagination:
        next_url = pagination['next']
        count+=1
    else:
        next_url = None

    if count > 91: 
        break
  
df = pd.DataFrame(all_data, columns=["id", "lat", "long"])
print(df)
print(len(df))
df.to_csv("testing.csv")
print(count)
```

result: 
```
                                           id        lat      long
0      https://api.weather.gov/stations/0007W  -84.17870  30.53099
1      https://api.weather.gov/stations/000PG -121.33943  36.78921
2      https://api.weather.gov/stations/000SE -117.85520  34.11814
3      https://api.weather.gov/stations/001HI -156.32000  20.75790
4      https://api.weather.gov/stations/001MD  -76.92580  39.26249
...                                       ...        ...       ...
45827  https://api.weather.gov/stations/ZPCC1 -120.29306  36.01639
45828  https://api.weather.gov/stations/ZSFO1  -82.39231  39.29158
45829  https://api.weather.gov/stations/ZTCN5 -107.21889  36.74972
45830  https://api.weather.gov/stations/ZVLT2  -94.38386  31.17720
45831  https://api.weather.gov/stations/ZZZI1 -115.97917  47.70750

[45832 rows x 3 columns]
45832
92
```

**Reply by DrewScatterday** [reply]
also there is tiny bug in this example code. The lat long points here should be flipped

**Reply by ftomasel** [reply]
Many thanks! There are now 63549 stations

**Reply by bcbeta** [reply]
> Many thanks! There are now 63549 stations

Only half of them have any data. It would be nice to be able to filter out the empty stations.

**Reply by ftomasel** [reply]
Indeed, only ~36k have any data - I am attaching the current list of 'active' stations for whatever is worth.  I agree - it would be good to have them somehow flagged in the metadata.
[stations.csv](https://github.com/user-attachments/files/22009090/stations.csv)

**Comment by branmh** [comment]
This page provides access to observed current weather conditions for about 1,800 locations across the United States and US Territories.

https://forecast.weather.gov/xml/index.xml

I wish the API would provide a list like this.
